index.php

```diff
@@ -158,22 +158,115 @@ function gutenberg_register_scripts() {
+		array( 'react' )
+	);
+	gutenberg_register_vendor_script(
+		'react-dom-server',
+		'https://unpkg.com/react-dom@next/umd/react-dom-server' . $react_suffix . '.js',
+		array( 'react' )
+	);
 
 	// Editor Scripts.
-	wp_register_script( 'tinymce-nightly', 'https://fiddle.azurewebsites.net/tinymce/nightly/tinymce' . $suffix . '.js' );
-	wp_register_script( 'wp-i18n', plugins_url( 'i18n/build/index.js', __FILE__ ), array(), filemtime( plugin_dir_path( __FILE__ ) . 'i18n/build/index.js' ) );
-	wp_register_script( 'wp-element', plugins_url( 'element/build/index.js', __FILE__ ), array( 'react', 'react-dom', 'react-dom-server' ), filemtime( plugin_dir_path( __FILE__ ) . 'element/build/index.js' ) );
-	wp_register_script( 'wp-blocks', plugins_url( 'blocks/build/index.js', __FILE__ ), array( 'wp-element', 'tinymce-nightly' ), filemtime( plugin_dir_path( __FILE__ ) . 'blocks/build/index.js' ) );
+	gutenberg_register_vendor_script(
+		'tinymce-nightly',
+		'https://fiddle.azurewebsites.net/tinymce/nightly/tinymce' . $suffix . '.js'
+	);
+	wp_register_script(
+		'wp-i18n',
+		plugins_url( 'i18n/build/index.js', __FILE__ ),
+		array(),
+		filemtime( plugin_dir_path( __FILE__ ) . 'i18n/build/index.js' )
+	);
+	wp_register_script(
+		'wp-element',
+		plugins_url( 'element/build/index.js', __FILE__ ),
+		array( 'react', 'react-dom', 'react-dom-server' ),
+		filemtime( plugin_dir_path( __FILE__ ) . 'element/build/index.js' )
+	);
+	wp_register_script(
+		'wp-blocks',
+		plugins_url( 'blocks/build/index.js', __FILE__ ),
+		array( 'wp-element', 'tinymce-nightly' ),
+		filemtime( plugin_dir_path( __FILE__ ) . 'blocks/build/index.js' )
+	);
 
 	// Editor Styles.
-	wp_register_style( 'wp-blocks', plugins_url( 'blocks/build/style.css', __FILE__ ), array(), filemtime( plugin_dir_path( __FILE__ ) . 'blocks/build/style.css' ) );
+	wp_register_style(
+		'wp-blocks',
+		plugins_url( 'blocks/build/style.css', __FILE__ ),
+		array(),
+		filemtime( plugin_dir_path( __FILE__ ) . 'blocks/build/style.css' )
+	);
 }
 add_action( 'init', 'gutenberg_register_scripts' );
 
 /**
+ * Registers a vendor script from a URL, preferring a locally cached version if
+ * possible, or downloading it if the cached version is unavailable or
+ * outdated.
+ *
+ * @param  string $handle Name of the script.
+ * @param  string $src    Full URL of the external script.
+ * @param  array  $deps   Optional. An array of registered script handles this
+ *                        script depends on.
+ *
+ * @since 0.1.0
+ */
+function gutenberg_register_vendor_script( $handle, $src, $deps = array() ) {
+	$filename = basename( $src );
+	$full_path = plugin_dir_path( __FILE__ ) . 'vendor/' . $filename;
+
+	$needs_fetch = (
+		! file_exists( $full_path ) ||
+		time() - filemtime( $full_path ) >= DAY_IN_SECONDS
+	);
+
+	if ( $needs_fetch ) {
+		// Determine whether we can write to this file.  If not, don't waste
+		// time doing a network request.
+		$f = fopen( $full_path, 'a' );
```

Comment:
Yes, I had originally done something like this, but `fopen( ..., 'w' )` [truncates the file](http://php.net/manual/en/function.fopen.php).

It might be possible to use `'r+'` but I wanted to avoid having to truncate the file and move the pointer around because this seemed pretty error-prone.